homeassistant/components/keyboard_remote/__init__.py

```diff
@@ -48,165 +60,210 @@
-                    KEYBOARD_REMOTE_DISCONNECTED,
-                    {
-                        DEVICE_DESCRIPTOR: self.device_descriptor,
-                        DEVICE_NAME: self.device_name,
-                    },
-                )
-                _LOGGER.debug("Keyboard disconnected, %s", self.device_id)
-                continue
+    class DeviceHandler:
+        """Manage input events using evdev with asyncio."""
 
-            if not event:
-                continue
+        def __init__(self, hass, dev_block):
+            """Fill configuration data."""
 
-            if event.type is ecodes.EV_KEY and event.value is self.key_value:
-                _LOGGER.debug(categorize(event))
-                self.hass.bus.fire(
-                    KEYBOARD_REMOTE_COMMAND_RECEIVED,
-                    {
-                        KEY_CODE: event.code,
-                        DEVICE_DESCRIPTOR: self.device_descriptor,
-                        DEVICE_NAME: self.device_name,
-                    },
-                )
+            self.hass = hass
 
+            key_types = dev_block.get(TYPE)
 
-class KeyboardRemote:
-    """Sets up one thread per device."""
+            self.key_values = set()
+            for key_type in key_types:
+                self.key_values.add(KEY_VALUE[key_type])
 
-    def __init__(self, hass, config):
-        """Construct a KeyboardRemote interface object."""
-        self.threads = []
-        for dev_block in config:
-            device_descriptor = dev_block.get(DEVICE_DESCRIPTOR)
-            device_name = dev_block.get(DEVICE_NAME)
-            key_value = KEY_VALUE.get(dev_block.get(TYPE, "key_up"))
+            self.emulate_key_hold = dev_block.get(EMULATE_KEY_HOLD)
+            self.emulate_key_hold_delay = dev_block.get(EMULATE_KEY_HOLD_DELAY)
+            self.emulate_key_hold_repeat = dev_block.get(EMULATE_KEY_HOLD_REPEAT)
+            self.monitor_task = None
+            self.dev = None
+
+        async def async_keyrepeat(self, dev, code, delay, repeat):
+            """Emulate keyboard delay/repeat behaviour by sending key events on a timer."""
 
-            if device_descriptor is not None or device_name is not None:
-                thread = KeyboardRemoteThread(
-                    hass, device_name, device_descriptor, key_value
+            await asyncio.sleep(delay)
+            while True:
+                self.hass.bus.async_fire(
+                    KEYBOARD_REMOTE_COMMAND_RECEIVED,
+                    {
+                        KEY_CODE: code,
+                        DEVICE_DESCRIPTOR: dev.path,
+                        DEVICE_NAME: dev.name,
+                    },
+                )
+                await asyncio.sleep(repeat)
+
+        async def async_start_monitoring(self, dev):
+            """Start event monitoring task and issue event."""
+            if self.monitor_task is None:
+                self.dev = dev
+                self.monitor_task = self.hass.loop.create_task(
```

Comment:
If we want home assistant to track the task we use `hass.async_create_task`. We don't want to track tasks that can block home assistant from finishing setup phase, before home assistant start event.

So normally we use `hass.async_create_task`, so that tasks have a chance to be cleaned up when stopping home assistant.

Reply:
Sorry I'm a bit unclear on the difference between hass.async_create_task and just calling loop.create_task.  Should I replace all the hass.loop.create_task calls with hass.async_create_task?

Reply:
Ok I figured this out.  Now all the tasks are started from separate functions called after EVENT_HOMEASSISTANT_START and consistently use hass.async_create_task